File: src/lib/ai/tasks/extract-candidate-profile.ts
import { AIResult } from "@/types";
import { callAI } from "../client";
import { callPythonAI, isPythonAIEnabled } from "../python-sidecar";
import {
  CandidateProfileImportSchema,
  validateAIOutput,
} from "../schemas";
import {
  CandidateProfileImportDraft,
  getDefaultCandidateProfile,
  normalizeCandidateProfile,
} from "@/lib/profile/candidate-profile";

const AI_CV_TEXT_LIMIT = 12_000;
const ROLE_TRACKS = [
  "qa",
  "regulatory",
  "pv",
  "medinfo",
  "clinical",
  "other",
] as const;

const SYSTEM_PROMPT = `You extract candidate profiles from CV text for a local-first career operating system.
Return valid JSON only.
Focus on UK life sciences transition roles and preserve specifics from the CV.
Do not invent certifications or achievements that are not grounded in the source text.`;

function buildPrompt(rawText: string, sourceFiles: string[]): string {
  const cvText = truncateForPrompt(rawText);

  return `Extract a normalized candidate profile draft from this CV text.

Return exactly:
{
  "rawText": "the original text summarized or truncated only if necessary",
  "profile": {
    "fullName": "string",
    "headline": "string",
    "location": "string",
    "openToRelocationUk": true,
    "summary": "string",
    "targetTitles": ["string"],
    "targetRoleTracks": ["clinical", "regulatory"],
    "locationConstraints": ["string"],
    "transitionNarrative": "string",
    "strengths": ["string"],
    "experienceHighlights": ["string"],
    "education": ["string"],
    "sourceCvIds": ["string"],
    "extraction": {
      "reviewState": "draft",
      "confidence": 0.0,
      "issues": ["string"],
      "extractedAt": "ISO string",
      "sourceFiles": ["string"]
    }
  },
  "confidence": 0.0,
  "issues": ["string"],
  "sourceFiles": ["string"],
  "extractedAt": "ISO string"
}

SOURCE FILES:
${sourceFiles.map((item) => `- ${item}`).join("\n")}

CV TEXT:
---
${cvText}
---

Respond with JSON only.`;
}

function truncateForPrompt(rawText: string): string {
  if (rawText.length <= AI_CV_TEXT_LIMIT) {
    return rawText;
  }

  return `${rawText.slice(0, AI_CV_TEXT_LIMIT)}

[CV text truncated for AI extraction. Preserve only facts visible above.]`;
}

export async function extractCandidateProfile(
  rawText: string,
  sourceFiles: string[]
): Promise<AIResult<CandidateProfileImportDraft> | { error: string }> {
  if (isPythonAIEnabled()) {
    try {
      const result = await callPythonAI<
        { rawText: string; sourceFiles: string[] },
        {
          success: boolean;
          data?: CandidateProfileImportDraft;
          meta?: AIResult<CandidateProfileImportDraft>["meta"];
          error?: string;
        }
      >(
        "/extract-candidate-profile",
        { rawText, sourceFiles },
        360_000
      );

      if (result.success && result.data && result.meta) {
        return {
          data: {
            ...result.data,
            profile: normalizeCandidateProfile(result.data.profile),
          },
          meta: result.meta,
          rawInput: { rawText, sourceFiles },
          rawOutput: "",
        };
      }

      console.warn(
        "[extract-candidate-profile] Python sidecar returned failure, falling back to TS:",
        result.error
      );
    } catch (err) {
      console.warn(
        "[extract-candidate-profile] Python sidecar call failed, falling back to TS:",
        err instanceof Error ? err.message : err
      );
    }
  }

  const result = await callAI<CandidateProfileImportDraft>({
    taskType: "extract-candidate-profile",
    prompt: buildPrompt(rawText, sourceFiles),
    systemPrompt: SYSTEM_PROMPT,
    rawInput: { rawText: truncateForPrompt(rawText), sourceFiles },
    temperature: 0.1,
  });

  if (!result.success || !result.data) {
    return buildHeuristicDraft(
      rawText,
      sourceFiles,
      result.error || "Candidate profile extraction failed"
    );
  }

  // Local 2B-7B Ollama models occasionally return a JSON array, a quoted
  // string, or markdown wrapping the object. Zod validation against
  // those shapes throws downstream errors (incl. the V8
  // "Object.defineProperty called on non-object" surface from the
  // candidate-profile UI). Reject non-object payloads up-front with a
  // clear message instead.
  const data: unknown = result.data;
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    return buildHeuristicDraft(
      rawText,
      sourceFiles,
      "AI returned a non-object payload for candidate profile extraction."
    );
  }

  const validation = validateAIOutput(
    CandidateProfileImportSchema,
    repairCandidateProfileImport(data)
  );
  if (!validation.valid) {
    return buildHeuristicDraft(
      rawText,
      sourceFiles,
      `AI validation failed: ${validation.error}`
    );
  }

  return {
    data: {
      ...validation.data,
      profile: normalizeCandidateProfile(validation.data.profile),
    },
    meta: {
      ...result.meta!,
      confidence: validation.data.confidence,
    },
    rawInput: { rawText, sourceFiles },
    rawOutput: result.rawOutput,
  };
}

function repairCandidateProfileImport(data: unknown): unknown {
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    return data;
  }

  const candidate = data as {
    profile?: {
      targetRoleTracks?: unknown;
    };
  };

  if (candidate.profile && typeof candidate.profile === "object") {
    candidate.profile = {
      ...candidate.profile,
      targetRoleTracks: coerceRoleTracks(candidate.profile.targetRoleTracks),
    };
  }

  return candidate;
}

function coerceRoleTracks(value: unknown): string[] {
  const inputs = Array.isArray(value) ? value : [value];
  const tracks = inputs.flatMap((item) =>
    typeof item === "string" ? item.split(/[|,/]/).map((part) => part.trim()) : []
  );
  const validTracks = tracks.filter((track): track is (typeof ROLE_TRACKS)[number] =>
    (ROLE_TRACKS as readonly string[]).includes(track)
  );

  return validTracks.length > 0 ? unique(validTracks) : ["other"];
}

function buildHeuristicDraft(
  rawText: string,
  sourceFiles: string[],
  reason: string
): AIResult<CandidateProfileImportDraft> {
  const extractedAt = new Date().toISOString();
  const defaults = getDefaultCandidateProfile();
  const lines = getUsefulLines(rawText);
  const fullName = inferFullName(lines) || defaults.fullName;
  const headline = inferHeadline(lines) || defaults.headline;
  const location = inferLocation(lines) || defaults.location;
  const education = inferEducation(lines, defaults.education);
  const strengths = inferStrengths(rawText, defaults.strengths);
  const experienceHighlights = inferExperienceHighlights(
    lines,
    defaults.experienceHighlights
  );

  const draft: CandidateProfileImportDraft = {
    rawText,
    profile: normalizeCandidateProfile({
      ...defaults,
      fullName,
      headline,
      location,
      summary: buildSummary(fullName, headline, location, education),
      strengths,
      experienceHighlights,
      education,
      sourceCvIds: sourceFiles,
      extraction: {
        reviewState: "draft",
        confidence: 0.35,
        issues: [
          `AI extraction did not complete cleanly, so this draft was created from basic CV text parsing. Original error: ${reason}`,
        ],
        extractedAt,
        sourceFiles,
      },
    }),
    confidence: 0.35,
    issues: [
      `AI extraction did not complete cleanly, so this draft needs review. Original error: ${reason}`,
    ],
    sourceFiles,
    extractedAt,
  };

  return {
    data: draft,
    meta: {
      model: "heuristic-fallback",
      promptType: "extract-candidate-profile",
      timestamp: extractedAt,
      confidence: draft.confidence,
      durationMs: 0,
      inputBytes: new TextEncoder().encode(rawText).length,
      outputBytes: 0,
      fallbackUsed: true,
      fallbackAttempted: true,
      attemptCount: 1,
      effectiveTimeoutMs: 0,
      jsonExtractionFallback: false,
    },
    rawInput: { rawText, sourceFiles },
    rawOutput: "",
  };
}

function getUsefulLines(rawText: string): string[] {
  return rawText
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length >= 2 && line.length <= 160)
    .slice(0, 80);
}

function inferFullName(lines: string[]): string {
  const candidate = lines.find(
    (line) =>
      /^[A-Z][A-Za-z'.-]+(?:\s+[A-Z][A-Za-z'.-]+){1,3}$/.test(line) &&
      !/(curriculum|vitae|resume|profile|contact|education|experience)/i.test(line)
  );
  return candidate || "";
}

function inferHeadline(lines: string[]): string {
  return (
    lines.find(
      (line) =>
        /clinical|research|trial|regulatory|quality|pharmacovigilance|medical information|CRA|CTA/i.test(
          line
        ) && !/@/.test(line)
    ) || ""
  );
}

function inferLocation(lines: string[]): string {
  return (
    lines.find((line) => /glasgow|scotland|london|uk|united kingdom|egypt/i.test(line)) ||
    ""
  );
}

function inferEducation(rawLines: string[], defaults: string[]): string[] {
  const education = rawLines.filter((line) =>
    /\b(MSc|BSc|PhD|Bachelor|Master|University|College|GCP)\b/i.test(line)
  );
  return unique([...education.slice(0, 8), ...defaults]);
}

function inferStrengths(rawText: string, defaults: string[]): string[] {
  const candidates = [
    ["GCP training", /\bGCP\b|good clinical practice/i],
    ["Clinical research coordination", /clinical research|trial coordination|CTA|CRA/i],
    ["Regulated healthcare documentation", /documentation|SOP|compliance|audit/i],
    ["Regulatory and quality awareness", /regulatory|quality|QA|GxP/i],
    ["Pharmacovigilance awareness", /pharmacovigilance|drug safety|PV/i],
  ] as const;

  return unique([
    ...candidates
      .filter(([, pattern]) => pattern.test(rawText))
      .map(([label]) => label),
    ...defaults,
  ]).slice(0, 10);
}

function inferExperienceHighlights(lines: string[], defaults: string[]): string[] {
  const highlights = lines.filter((line) =>
    /intern|assistant|coordinator|research|clinical|hospital|pharma|laboratory|trial/i.test(
      line
    )
  );
  return unique([...highlights.slice(0, 8), ...defaults]).slice(0, 10);
}

function buildSummary(
  fullName: string,
  headline: string,
  location: string,
  education: string[]
): string {
  const subject = fullName || "Candidate";
  const headlineText = headline
    ? ` with CV evidence around ${headline}`
    : " with CV evidence for regulated life sciences transition roles";
  const locationText = location ? ` Based in or connected to ${location}.` : "";
  const educationText = education.length > 0 ? ` Education/training: ${education[0]}.` : "";
  return `${subject}${headlineText}.${locationText}${educationText}`.trim();
}

function unique(values: string[]): string[] {
  return Array.from(
    new Set(values.map((value) => value.trim()).filter(Boolean))
  );
}

File: python-ai/life_os_ai/tasks/extract_candidate_profile.py
"""Extract a candidate profile draft from CV text.

This mirrors `src/lib/ai/tasks/extract-candidate-profile.ts`, but keeps
the parsing and repair logic in Python so the local app can gradually move
AI/data-shaping work out of the Next.js process.
"""

from __future__ import annotations

import json
import logging
import re
import time
from datetime import UTC, datetime
from typing import Any

from pydantic import ValidationError

from ..llm import chat
from ..schemas import AIMetadata, CandidateProfileImportDraft

log = logging.getLogger(__name__)

ROLE_TRACKS = ("qa", "regulatory", "pv", "medinfo", "clinical", "other")
AI_CV_TEXT_LIMIT = 12_000

DEFAULT_TARGET_TITLES = [
    "Clinical Trial Assistant",
    "Clinical Research Coordinator",
    "Clinical Operations Assistant",
    "Clinical Study Coordinator",
    "Trial Administrator",
    "Junior CRA",
    "QA Associate",
    "Quality Systems Associate",
    "Document Control Associate",
    "Regulatory Affairs Assistant",
    "Pharmacovigilance Associate",
    "Drug Safety Associate",
    "Medical Information Associate",
]
DEFAULT_TARGET_TRACKS = ["qa", "regulatory", "pv", "medinfo", "clinical"]
DEFAULT_LOCATIONS = ["Glasgow", "Scotland", "United Kingdom", "Remote", "Hybrid"]
DEFAULT_STRENGTHS = [
    "GCP training",
    "Regulated healthcare documentation",
    "SOP and compliance-heavy workflows",
    "Trial support and coordination readiness",
]

SYSTEM_PROMPT = (
    "You extract candidate profiles from CV text for a local-first career operating system. "
    "Return valid JSON only. Preserve specifics from the CV. Do not invent certifications "
    "or achievements that are not grounded in the source text."
)


def _truncate(raw_text: str) -> str:
    if len(raw_text) <= AI_CV_TEXT_LIMIT:
        return raw_text
    return (
        raw_text[:AI_CV_TEXT_LIMIT]
        + "\n\n[CV text truncated for AI extraction. Preserve only facts visible above.]"
    )


def _build_prompt(raw_text: str, source_files: list[str]) -> str:
    files = "\n".join(f"- {item}" for item in source_files) or "- unknown"
    return f"""\
Extract a normalized candidate profile draft from this CV text.

Return exactly this JSON shape. For targetRoleTracks, use only separate values from:
["qa", "regulatory", "pv", "medinfo", "clinical", "other"].

{{
  "rawText": "the original text summarized or truncated only if necessary",
  "profile": {{
    "fullName": "string",
    "headline": "string",
    "location": "string",
    "openToRelocationUk": true,
    "summary": "string",
    "targetTitles": ["string"],
    "targetRoleTracks": ["clinical", "regulatory"],
    "locationConstraints": ["string"],
    "transitionNarrative": "string",
    "strengths": ["string"],
    "experienceHighlights": ["string"],
    "education": ["string"],
    "sourceCvIds": ["string"],
    "extraction": {{
      "reviewState": "draft",
      "confidence": 0.0,
      "issues": ["string"],
      "extractedAt": "ISO string",
      "sourceFiles": ["string"]
    }}
  }},
  "confidence": 0.0,
  "issues": ["string"],
  "sourceFiles": ["string"],
  "extractedAt": "ISO string"
}}

SOURCE FILES:
{files}

CV TEXT:
---
{_truncate(raw_text)}
---

Respond with JSON only."""


def _extract_json(raw: str) -> dict[str, Any] | None:
    text = raw.strip()
    try:
        parsed = json.loads(text)
        return parsed if isinstance(parsed, dict) else None
    except json.JSONDecodeError:
        pass

    fence = re.search(r"```(?:json)?\s*([\s\S]*?)\s*```", text)
    if fence:
        try:
            parsed = json.loads(fence.group(1))
            return parsed if isinstance(parsed, dict) else None
        except json.JSONDecodeError:
            pass

    obj = re.search(r"\{[\s\S]*\}", text)
    if obj:
        try:
            parsed = json.loads(obj.group(0))
            return parsed if isinstance(parsed, dict) else None
        except json.JSONDecodeError:
            pass

    return None


def _unique(values: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for value in values:
        cleaned = value.strip()
        if cleaned and cleaned not in seen:
            seen.add(cleaned)
            result.append(cleaned)
    return result


def _coerce_role_tracks(value: Any) -> list[str]:
    values = value if isinstance(value, list) else [value]
    tracks: list[str] = []
    for item in values:
        if not isinstance(item, str):
            continue
        for part in re.split(r"[|,/]", item):
            cleaned = part.strip()
            if cleaned in ROLE_TRACKS:
                tracks.append(cleaned)
    return _unique(tracks) or ["other"]


def _repair(data: dict[str, Any], raw_text: str, source_files: list[str]) -> dict[str, Any]:
    extracted_at = datetime.now(UTC).isoformat()
    profile = data.get("profile") if isinstance(data.get("profile"), dict) else {}
    heuristic = _heuristic_profile(raw_text, source_files, "AI output was repaired before validation")
    heuristic_profile = heuristic["profile"]

    repaired_profile = {
        **heuristic_profile,
        **profile,
        "targetRoleTracks": _coerce_role_tracks(profile.get("targetRoleTracks")),
        "targetTitles": _string_list(profile.get("targetTitles")) or heuristic_profile["targetTitles"],
        "locationConstraints": _string_list(profile.get("locationConstraints"))
        or heuristic_profile["locationConstraints"],
        "strengths": _string_list(profile.get("strengths")) or heuristic_profile["strengths"],
        "experienceHighlights": _string_list(profile.get("experienceHighlights"))
        or heuristic_profile["experienceHighlights"],
        "education": _string_list(profile.get("education")) or heuristic_profile["education"],
        "sourceCvIds": _string_list(profile.get("sourceCvIds")) or source_files,
    }
    repaired_profile["extraction"] = {
        "reviewState": "draft",
        "confidence": _number(profile.get("extraction", {}).get("confidence"), 0.75)
        if isinstance(profile.get("extraction"), dict)
        else 0.75,
        "issues": _string_list(profile.get("extraction", {}).get("issues"))
        if isinstance(profile.get("extraction"), dict)
        else [],
        "extractedAt": extracted_at,
        "sourceFiles": source_files,
    }

    return {
        "rawText": data.get("rawText") if isinstance(data.get("rawText"), str) else raw_text,
        "profile": repaired_profile,
        "confidence": _number(data.get("confidence"), 0.75),
        "issues": _string_list(data.get("issues")),
        "sourceFiles": _string_list(data.get("sourceFiles")) or source_files,
        "extractedAt": data.get("extractedAt")
        if isinstance(data.get("extractedAt"), str)
        else extracted_at,
    }


def _string_list(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [item.strip() for item in value if isinstance(item, str) and item.strip()]


def _number(value: Any, fallback: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return fallback
    return max(0.0, min(1.0, number))


def _useful_lines(raw_text: str) -> list[str]:
    return [
        line.strip()
        for line in raw_text.splitlines()
        if 2 <= len(line.strip()) <= 160
    ][:80]


def _infer_full_name(lines: list[str]) -> str:
    for line in lines[:10]:
        if re.match(r"^[A-Z][A-Za-z'.-]+(?:\s+[A-Z][A-Za-z'.-]+){1,3}$", line) and not re.search(
            r"curriculum|vitae|resume|profile|contact|education|experience", line, re.I
        ):
            return line
    return ""


def _infer_headline(lines: list[str]) -> str:
    for line in lines:
        if re.search(
            r"clinical|research|trial|regulatory|quality|pharmacovigilance|medical information|CRA|CTA",
            line,
            re.I,
        ) and "@" not in line:
            return line
    return "Entry-level clinical operations, QA, regulatory, PV and medinfo candidate"


def _infer_location(lines: list[str]) -> str:
    for line in lines:
        if re.search(r"glasgow|scotland|london|uk|united kingdom|egypt", line, re.I):
            return line
    return "Glasgow, Scotland"


def _infer_education(lines: list[str]) -> list[str]:
    found = [
        line
        for line in lines
        if re.search(r"\b(MSc|BSc|PhD|Bachelor|Master|University|College|GCP)\b", line, re.I)
    ][:8]
    return _unique(found + ["MSc Clinical Pharmacology"])


def _infer_strengths(raw_text: str) -> list[str]:
    candidates = [
        ("GCP training", r"\bGCP\b|good clinical practice"),
        ("Clinical research coordination", r"clinical research|trial coordination|CTA|CRA"),
        ("Regulated healthcare documentation", r"documentation|SOP|compliance|audit"),
        ("Regulatory and quality awareness", r"regulatory|quality|QA|GxP"),
        ("Pharmacovigilance awareness", r"pharmacovigilance|drug safety|PV"),
    ]
    found = [label for label, pattern in candidates if re.search(pattern, raw_text, re.I)]
    return _unique(found + DEFAULT_STRENGTHS)[:10]


def _infer_experience(lines: list[str]) -> list[str]:
    found = [
        line
        for line in lines
        if re.search(
            r"intern|assistant|coordinator|research|clinical|hospital|pharma|laboratory|trial",
            line,
            re.I,
        )
    ][:8]
    fallback = [
        "Clinical research internship exposure",
        "Healthcare administration and regulated support workflows",
        "Documentation quality, filing, archiving and audit-readiness mindset",
    ]
    return _unique(found + fallback)[:10]


def _heuristic_profile(raw_text: str, source_files: list[str], reason: str) -> dict[str, Any]:
    extracted_at = datetime.now(UTC).isoformat()
    lines = _useful_lines(raw_text)
    full_name = _infer_full_name(lines)
    headline = _infer_headline(lines)
    location = _infer_location(lines)
    education = _infer_education(lines)
    strengths = _infer_strengths(raw_text)
    experience = _infer_experience(lines)
    summary_subject = full_name or "Candidate"

    return {
        "rawText": raw_text,
        "profile": {
            "fullName": full_name,
            "headline": headline,
            "location": location,
            "openToRelocationUk": True,
            "summary": (
                f"{summary_subject} with CV evidence around {headline}. "
                f"Based in or connected to {location}. Education/training: {education[0]}."
            ),
            "targetTitles": DEFAULT_TARGET_TITLES,
            "targetRoleTracks": DEFAULT_TARGET_TRACKS,
            "locationConstraints": DEFAULT_LOCATIONS,
            "transitionNarrative": (
                "Targeting desk-based transition roles in UK life sciences and pharma, "
                "with focus on clinical operations, QA, regulatory, PV, and medinfo pathways."
            ),
            "strengths": strengths,
            "experienceHighlights": experience,
            "education": education,
            "sourceCvIds": source_files,
            "extraction": {
                "reviewState": "draft",
                "confidence": 0.35,
                "issues": [
                    "AI extraction did not complete cleanly, so this draft was created "
                    f"from basic CV text parsing. Original error: {reason}"
                ],
                "extractedAt": extracted_at,
                "sourceFiles": source_files,
            },
        },
        "confidence": 0.35,
        "issues": [
            "AI extraction did not complete cleanly, so this draft needs review. "
            f"Original error: {reason}"
        ],
        "sourceFiles": source_files,
        "extractedAt": extracted_at,
    }


def extract_candidate_profile(
    raw_text: str, source_files: list[str]
) -> tuple[CandidateProfileImportDraft, AIMetadata]:
    started = time.time()
    input_bytes = len(raw_text.encode("utf-8"))

    result = chat(
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": _build_prompt(raw_text, source_files)},
        ],
        temperature=0.1,
        max_tokens=1600,
        response_format="json",
    )
    duration_ms = int((time.time() - started) * 1000)

    def _fallback(kind: str, reason: str) -> tuple[CandidateProfileImportDraft, AIMetadata]:
        draft = CandidateProfileImportDraft.model_validate(
            _heuristic_profile(raw_text, source_files, reason)
        )
        meta = AIMetadata.model_validate(
            {
                "model": "heuristic-fallback",
                "promptType": "extract-candidate-profile",
                "timestamp": datetime.now(UTC).isoformat(),
                "confidence": draft.confidence,
                "durationMs": duration_ms,
                "inputBytes": input_bytes,
                "outputBytes": len(draft.model_dump_json().encode("utf-8")),
                "fallbackUsed": True,
                "fallbackAttempted": result.fallback_used,
                "attemptCount": 1,
                "effectiveTimeoutMs": 0,
                "jsonExtractionFallback": False,
                "failureKind": kind,
            }
        )
        return draft, meta

    if not result.success or result.text is None:
        reason = result.error or "Candidate profile extraction failed"
        log.warning("extract-candidate-profile LLM failed (%s)", result.failure_kind)
        return _fallback(result.failure_kind or "runtime_error", reason)

    data = _extract_json(result.text)
    if data is None:
        return _fallback("invalid_json", "Could not extract valid JSON from AI output")

    try:
        draft = CandidateProfileImportDraft.model_validate(_repair(data, raw_text, source_files))
    except ValidationError as exc:
        log.warning("candidate profile validation failed after repair: %s", exc)
        return _fallback("schema_validation", f"AI validation failed: {exc}")

    meta = AIMetadata.model_validate(
        {
            "model": result.model or "unknown",
            "promptType": "extract-candidate-profile",
            "timestamp": datetime.now(UTC).isoformat(),
            "confidence": draft.confidence,
            "durationMs": result.duration_ms,
            "inputBytes": input_bytes,
            "outputBytes": len(result.text.encode("utf-8")),
            "fallbackUsed": result.fallback_used,
            "fallbackAttempted": result.fallback_used,
            "attemptCount": 1,
            "effectiveTimeoutMs": 0,
            "jsonExtractionFallback": False,
        }
    )
    return draft, meta
